src/main/java/org/wildfly/security/authz/MappedRoleMapper.java

```diff
@@ -0,0 +1,171 @@
+
+            for (String mappedRole : entry.getValue()) {
+                Set<String> rolesToMappedRole = reverseRoleMap.get(mappedRole);
+
+                if (rolesToMappedRole == null) {
+                    rolesToMappedRole = new LinkedHashSet<>();
+                    reverseRoleMap.put(mappedRole, rolesToMappedRole);
+                }
+
+                rolesToMappedRole.add(entry.getKey());
+            }
+        }
+    }
+
+
+    /**
+     * Custom component method.
+     * Called from WildFly core. Used to include mapped role mapping functionality in older WildFly versions.
+     *
+     * @param configuration map of mapping rules
+     */
+    public void initialize(final Map<String, String> configuration) {
+        configuration.forEach( (key, value) -> {
+            String[] newRoles = value.split(",");
+            for (String newRole : newRoles) {
+                Set<String> rolesToMappedRole = reverseRoleMap.get(newRole);
+                if (rolesToMappedRole == null) {
+                    rolesToMappedRole = new LinkedHashSet<>();
+                    reverseRoleMap.put(newRole, rolesToMappedRole);
+                }
+                rolesToMappedRole.add(value);
+            }
+        } );
+    }
+
+    @Override
+    public Roles mapRoles(Roles rolesToMap) {
+        Roles res = new MappedRoles(rolesToMap, this.reverseRoleMap);
+        if (!removeMapped) {
+            res = res.or(rolesToMap.and(keyRoles));
+        }
+        if (!removeNonMapped) {
+            res = res.or(rolesToMap.minus(keyRoles));
+        }
+        return res;
+    }
+
+    /**
+     * Construct a new {@link Builder} for creating the {@link MappedRoleMapper}.
+     *
+     * @return a new {@link Builder} for creating the {@link MappedRoleMapper}.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * A builder for map backed role mappers.
+     */
+    public static class Builder {
+        private Map<String, Set<String>> roleMap;
+        private boolean removeMapped = true;
+        private boolean removeNonMapped = true;
+
+        Builder() {
+        }
+
+        /**
+         * Build and return the resulting {@link MappedRoleMapper}.
+         *
+         * @return the resulting {@link MappedRoleMapper}
+         */
+        public MappedRoleMapper build() {
+            return new MappedRoleMapper(roleMap, removeMapped, removeNonMapped);
+        }
+
+        /**
+         * Set the mapper to remove roles, that were changed by its {@link Map}
+         *
+         * @param removeMapped true if original mapped roles should be removed
+         * @return {@code this} builder to allow chaining.
+         */
+        public Builder setRemoveMapped(boolean removeMapped) {
+            this.removeMapped = removeMapped;
+            return this;
+        }
+
+        /**
+         * Set the mapper to remove roles, that were not changed by its {@link Map}
```

Comment:
"to remove roles without defined mapping instead of keeping them" ?
Maybe would be better to describe opposite meaning of the booleans - setKeepNonmapped/setKeepOriginal?